### ISSUE TITLE
feat(plugin-roles): new plugin for assigning roles

### ISSUE THREAD

**Issue by Pkmmte** [Wave-Play/robo.js#134]
Changeset: `minor`

**Comment by ArnavK-09** [comment]
what would be use of list? like people can choose roles from it?

**Comment by Pkmmte** [comment]
> what would be use of list? like people can choose roles from it?

Yes. It'd be one of those common "choose an emote to get the role" kind of thing. I figured it's such a common use case that we might as well support it officially!

**Comment by ArnavK-09** [comment]
## Personal Todo

#### role setup
- [x] cmd init & flashcore init
- [ ] add member check before interactions for plugin
- [ ] fix regexps
- [ ] add comments
- [ ] add variable names
- [ ] check working
- [x] add roles drop-down in setup msg
- [x] single role per interaction
- [ ] show modal for emote & description
- [x] realtime embed review
- [ ] delete role button that triggers modal  asking role index
- [ ] custom embed modal primary btn
- [x] print setup embed & drop-down
- [ ] cleanup delete db setup instance
- [ ] add timeout if possible
- [ ] admin check 
- [x] role POS check
- [ ] good ux err msgs


#### role restrict
- [x] init
- [ ] admin check
- [ ] autocomplete with portal cmds
- [ ] role opt in cmd Slash
- [ ] confirmation
- [ ] code qc
- [ ] testing
- [ ] adminstrator bypass
- [ ] middleware file abort
- [ ] good ux replies 


> !Ignore

**Comment by Pkmmte** [comment]
Alright, so I just checked it out and found a few things to note:

- Buttons for "Add", "Edit", and "Delete" are missing. These should show alongside the "Publish" button already present.
<img width="747" alt="Screenshot 2023-12-11 at 10 16 49 PM" src="https://github.com/Wave-Play/robo.js/assets/3953360/7f6286b1-81a7-4035-9e46-2ed839bafd90">


- The "Edit Embed" is a good replacement for the "Description" outlined. I like it! Let's keep it.
<img width="456" alt="Screenshot 2023-12-11 at 10 18 11 PM" src="https://github.com/Wave-Play/robo.js/assets/3953360/df52546c-01f1-4a06-8513-dad03aee439f">

- There should not be selectors in the primary `/roles setup` present. Currently there are two selectors present. The "Add" and "Delete" buttons are meant to handle this instead. More on this below.
<img width="438" alt="Screenshot 2023-12-11 at 10 18 45 PM" src="https://github.com/Wave-Play/robo.js/assets/3953360/d2cfbf73-f278-408f-aa92-3a9b0e5098d4">

- We should be able to use even custom server emotes, not just default system emotes. This is very important. Since you brought up the concern about selectors having a 25 choice limit, let's do this instead: 
  - Clicking on "Add" should have the bot reply with a non-ephemeral message (due to reaction limits). This message should contain a role selector to choose only one role, along with instructions telling the user to react to the message with an emote to use for the role. This can be done either before or after the role selector is used.
  - The idea is that admins/mods would use this in a private channel, but let's add safety measures anyway in case it's done publicly. Make sure only the person that clicked on "Add" can react or use the role selector. For reactions, if the user is not the same person that triggered it, the reaction should be removed.
  - Once a role is selected and a reaction is added, the bot should apply these to the saved emotes and auto delete the message they just reacted on. If there's no activity for 10 minutes, the message gets removed anyway to prevent spam.

- The "Delete" button should also create a message similar to "Add". Either selecting a role or reacting with an emote removes that entry and causes the message to self-destruct.

- If you'd like, you can remove the "Edit" button altogether since the same can be achieved by "Delete" followed by "Add" anyway.

- "Publish" has a small typo. Currently says "Publush Setup!".

- When the "Publish" button is pressed, the bot should reply with an ephemeral message containing a channel selector. This allows an admin/moderator to do the entire setup in a private channel while publishing it to a public channel of their choice.

- Server users should be able to pick a role by adding a reaction, not using a selector. This is an important requirement as it follows an already-established pattern.

Sorry if I wasn't more clear in the GitHub Issue. I look forward to seeing this fully completed. We've been getting more traffic to our blogs lately so I'd love to write an article about this plugin!

**Comment by ArnavK-09** [comment]
> When the "Publish" button is pressed, the bot should reply with an ephemeral message containing a channel selector. This allows an admin/moderator to do the entire setup in a private channel while publishing it to a public channel of their choice

Noted, would update soon after we solve add debate
